File: Codesnippets/Successor/Iterative/successor.cpp
#include <iostream>
#include <string>

/* * Iterative successor function.
 * Modifies the string in place and returns true if an overflow occurred.
 */
bool bitstring_successor_iterative(std::string& bits){
  if (bits.empty()){
    return false;
  }

  // Start at the rightmost bit and move left
  for (int i = bits.length() - 1; i >= 0; i--) {
    if (bits[i] == '1') {
      // The bit is 1, so 1 + 1 = 0 with a carry of 1.
      // We set it to 0 and let the loop continue to the next bit.
      bits[i] = '0';
    } else {
      // We found the first '0'. 0 + 1 (carry) = 1;
      // The carry is now consumed (0), so we can stop immediately!
      bits[i] = '1';
      return false; // No global overflow
    }
  }

  // If the loop finishes completely without hitting a 'return false',
  // it means every single bit was a '1' (e.g., "1111").
  return true; // Global overflow occurred
}

int main() {
  // Test case 1: Standard iterative increment (11 -> 12)
  std::string binary_sample1 = "1011";

  std::cout << "Before: " << binary_sample1 << std::endl;
  bool overflow1 = bitstring_successor_iterative(binary_sample1);
  std::cout << "After: " << binary_sample1 << " (Overflow: " << (overflow1 ? "Yes" : "No") << ")\n\n";

  // Test case 2: Edge case with global overflow (7 -> 0)
  std::string binary_sample2 = "111";

  std::cout << "Before: " << binary_sample2 << std::endl;
  bool overflow2 = bitstring_successor_iterative(binary_sample2);
  std::cout << "After: " << binary_sample2 << " (Overflow: " << (overflow2 ? "Yes" : "No") << ")\n\n";

  return 0;
}

File: Codesnippets/Successor/Iterative/successor.py
def bitstring_successor_iterative(bits_str):
  if not bits_str:
    return False, bits_str

  # Convert the string into a mutable list of characters
  char_list = list(bits_str)

  # Iterate from right to the left: range(start, stop_exclusive, step)
  for i in range(len(char_list) -1, -1, -1):
    if char_list[i] == '1':
      char_list[i] = '0' # 1 + 1 = 0 (Carry moves further to the left)
    else:
      char_list[i] = '1' # First '0' found, carry is consumed

      # Reconstruct the string and trigger early exit (no global overflow)
      return False, "".join(char_list)

  #If the loop finishes completely, all bits were '1' (e.g., "111")
  return True, "".join(char_list)

#===============================================
#=============== TEST CASE =====================
#===============================================
if __name__ == "__main__":
  # Test case 1: Standard iterative increment (11 -> 12)
  sample_1 = "1011"
  print(f"Before: {sample_1}")
  overflow_1, result_1 = bitstring_successor_iterative(sample_1)
  print(f"After: {result_1} (Overflow: {'Yes' if overflow_1 else 'No'})\n")

  # Test case 2: Edge case with global overflow (7 -> 0)
  sample_2 = "111"
  print(f"Before: {sample_2}")
  overflow_2, result_2 = bitstring_successor_iterative(sample_2)
  print(f"After: {result_2} (Overflow: {'Yes' if overflow_2 else 'No'})")
  
